File: day-02/ruby/main.rb
#! /usr/bin/env ruby

require_relative 'puzzel'

sample = Puzzel::Sample.new.values
input  = Puzzel::Input.new.values

HANDS = ["Rock", "Paper", "Scissors"]
TYPE  = ["Loss", "Draw", "Win"]

module Game

  module Support

    def hand
      self.class.name.split('::').last
    end

  end

  class Play

    attr_reader :result

    def initialize(player_1, player_2)
      @player_1 = player_1
      @player_2 = player_2
      @result   = run
    end

    private

    def run
      case
      when win?
        Result::Win.new(@player_2)
      when draw?
        Result::Draw.new(@player_2)
      else
        Result::Loss.new(@player_2)
      end
    end

    def win?
      @player_2.beats == @player_1.hand
    end

    def draw?
      @player_2.hand == @player_1.hand
    end

  end

  module Result

    class Type
      include Support

      def initialize(play)
        @play = play
      end

      def score
        Hash[TYPE.zip([0,3,6])][hand]
      end

      def total
        score + @play.score
      end

    end

    class Loss < Type
    end

    class Draw < Type
    end

    class Win < Type
    end

  end

  module RPS

    class Hand
      include Support

      def score
        Hash[HANDS.zip([1,2,3])][hand]
      end

      def beats
        Hash[HANDS.zip(HANDS.rotate(2))][hand]
      end

    end

    class Rock < Hand
    end

    class Paper < Hand
    end

    class Scissors < Hand
    end

  end

end

hands = [Game::RPS::Rock.new, Game::RPS::Paper.new, Game::RPS::Scissors.new]

set_1 = ["A", "B", "C"]
set_2 = ["X", "Y", "Z"]

# # Part 1 Sample

results = sample.map do |players|
  Game::Play.new(
    *players.split(" ").map do |play| 
      hands[(set_1.index(play) || set_2.index(play))].clone
    end
  ).result
end

printf(
  "Day 02: Part 1: [Sample]: Total Score: %6d\n", 
  results.collect(&:total).inject(&:+)
)

# # Part 1 Input

results = input.map do |players|
  Game::Play.new(
    *players.split(" ").map do |play| 
      hands[(set_1.index(play) || set_2.index(play))].clone
    end
  ).result
end

printf(
  "Day 02: Part 1:  [Input]: Total Score: %6d\n", 
  results.collect(&:total).inject(&:+)
)

# # Part 2 Sample

set_2 = set_2.rotate(1)

results = sample.map do |players|
  p1, p2 = players.split(" ").map {|p0| (set_1.index(p0) || set_2.index(p0)) }
  Game::Play.new( hands[p1].clone, hands.rotate(p2)[p1].clone ).result
end

printf(
  "Day 02: Part 2: [Sample]: Total Score: %6d\n",
  results.collect(&:total).inject(&:+)
)

# # Part 2 Input

results = input.map do |players|
  p1, p2 = players.split(" ").map {|p0| (set_1.index(p0) || set_2.index(p0)) }
  Game::Play.new( hands[p1].clone, hands.rotate(p2)[p1].clone ).result
end

printf(
  "Day 02: Part 2:  [Input]: Total Score: %6d\n",
  results.collect(&:total).inject(&:+)
)

File: day-02/python/main.py
#! /usr/bin/env python

import Sample
import Input
import numpy as np

sample = Sample.Puzzel()
input  = Input.Puzzel()

HANDS = ["Rock", "Paper", "Scissors"]
TYPE  = ["Loss", "Draw", "Win"]

class Hand:
  def __init__(self):
    self.hand = str(self.__class__.__name__)

  def score(self):
    return { key: value for key, value in zip(HANDS, [1,2,3]) }[self.hand]

  def beats(self):
    return { key: value for key, value in zip(HANDS, np.roll(HANDS, 1)) }[self.hand]

class Rock(Hand):
  pass

class Paper(Hand):
  pass

class Scissors(Hand):
  pass

class ResultType:
  def __init__(self, play):
    self.play = play
    self.type = self.__class__.__name__

  def score(self):
    return { key: value for key, value in zip(TYPE, [0,3,6]) }[self.type]

  def total(self):
    return self.score() + self.play.score()

class Loss(ResultType):
  pass

class Draw(ResultType):
  pass

class Win(ResultType):
  pass

class Play:
  def __init__(self, player_1, player_2):
    self.player_1 = player_1
    self.player_2 = player_2
    self.result   = self.__run()

  def __run(self):
    if self.__is_win():
      return Win(self.player_2)
    elif self.__is_draw():
      return Draw(self.player_2)
    else:
      return Loss(self.player_2)

  def __is_win(self):
    return self.player_2.beats() == self.player_1.hand

  def __is_draw(self):
    return self.player_2.hand == self.player_1.hand

set_1 = ["A", "B", "C"]
set_2 = ["X", "Y", "Z"]

def index(value):
  if value in set_1: return set_1.index(value)
  if value in set_2: return set_2.index(value)

hands = [Rock(), Paper(), Scissors()]

# # Part 1
# Sample
results = [Play(*[hands[index(play)] for play in plays.split()]).result for plays in sample.values()]

print(
  "Day 02: Part 1: [Sample]: Total Score: ", 
  str(sum([result.total() for result in results]))
)

# Input
results = [Play(*[hands[index(play)] for play in plays.split()]).result for plays in input.values()]

print(
  "Day 02: Part 1:  [Input]: Total Score: ", 
  str(sum([result.total() for result in results]))
)

# # Part 2
# Sample
set_2 = list(np.roll(set_2, -1))

results = []

for plays in sample.values():
  pl = [index(ply) for ply in plays.split(" ")]
  results.append(Play(hands[pl[0]], np.roll(hands, -pl[1])[pl[0]]).result)

print(
  "Day 02: Part 1: [Sample]: Total Score: ", 
  str(sum([result.total() for result in results]))
)

# Input
results = []

for plays in input.values():
  pl = [index(ply) for ply in plays.split(" ")]
  results.append(Play(hands[pl[0]], np.roll(hands, -pl[1])[pl[0]]).result)

print(
  "Day 02: Part 1:  [Input]: Total Score: ", 
  str(sum([result.total() for result in results]))
)